[WIP] [FEATURE] [needs-docs] Fixes #16047 - Use raster band names in QGIS

AndrewAnnex opened the pull request:
## Description
Fixes #16047
QGIS currently does not utilize band names present in gdal datasets within the UI except via a specialized metadata parser currently within qgsgdalprovider.cpp. Hyperspectral datasets or those with dozens of bands with descriptive band names are currently difficult to work with in QGIS as one has to refer to a look up table despite the meta data being present in the file. 

This WIP PR changes this by providing configurable behavior to optionally read out this metadata from the gdal description.

`Data source handling` now has five options for behavior to present band names within the ui, with the default case essentially being the current behavior.
1. "Band Number" -> returns band number between 1 and N bands.
2. "Band Name" -> returns gdal band description
3. "Band Name & Number" -> returns gdal band description with number appended, e.g. `nvdi (Band 1)`
4. "Band Number & Name" -> reverse of 3rd option, e.g. `Band 1 (nvdi)`
5. "Use _DIM_EXTRA" -> Specialized code path to support those users

This PR is a WIP currently because I am looking for feedback, but we could promote it to a full pr if scope is correct. I have not fully investigated changing the behavior elsewhere i.e. in the raster histogram panel view or in the raster calculator, and I could imagine that there are cases for generic raster layers that I am not sure if are necessary to address to fix the issue. 

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

AndrewAnnex committed:
Use raster band names in QGIS [FEATURE] [needs-docs]
* fixes #16047
* extracts band name info using gdalapi in QgsGdalProvider
* added configuration option to allowing user to select band name behavior
* currently 5 ways including current default behavior: #, name, name (#), # (name), _DIM_EXTRA metadata

luipir commented:
@AndrewAnnex do you mind to add tests for this feature?

nyalldawson commented:
Nice work!

My initial thoughts are whether or not we can avoid the extra option here, and just choose a default behaviour which addresses the most common situations. Options add UI complexity and extra code maintenance burden, so ideally we can find a standard behaviour to use without exposing this choice. What do you think?

AndrewAnnex commented:
@luipir will do, but it will probably be a week before I get a chance to

@nyalldawson well that is definitely a question I had, on balance I think `band number (band name)` ie `Band 01 (NVDI)` works as a good default, so if a raster lacks descriptors on the bands it would just fall back to `Band 01` for that example. But I can imagine arguments being made for dropping the number if the name is available or for making it configurable to satisfy everyone. As I need to do other things for about a week hopefully other maintainers can comment on this

@seignovert Hello again! This pr does some work in the same area as your pr #7317 , do you have opinions on these changes?

nyalldawson commented:
> well that is definitely a question I had, on balance I think band number (band name) ie Band 01 (NVDI) works as a good default, so if a raster lacks descriptors on the bands it would just fall back to Band 01 for that example. But I can imagine arguments being made for dropping the number if the name is available or for making it configurable to satisfy everyone.

Go with your gut here -- I think `band number (band name)` (or `band number: band name`, or `band number -  band name`) is a good all round option.

If you do this, and drop the option, it classifies as a bug fix and can be merged for 3.4. 

> But I can imagine arguments being made for dropping the number if the name is available or for making it configurable to satisfy everyone

We can't please everyone, and aiming to do so will just make QGIS an unusable mess of a 1000 different options :) In 3.0 we actually dropped a bunch of options for this very reason, and I'd be opposed to seeing unnecessary options creep back in again...

AndrewAnnex committed:
responding to github comments, simplified behavior to display the band name if present 'Band #: Band Name'. removed options

AndrewAnnex committed:
added unit test and test raster

AndrewAnnex committed:
forgot to add voids in testqgsgdalprovider.cpp

AndrewAnnex commented:
hey all, so I simplified things greatly by removing the configuration bits. I added two tests and a very small test raster (328B) for testqgsgdalprovider. I currently have the behavior set to use colon space for the band name (`Band 1: NVDI`) but I noticed that the code for the `DIM_EXTRA`  code uses a space and slash (`Band 1 / NVDI`). We should probably do one or the other though, I favor using the colon as it is a little more compact.

nyalldawson commented:
@AndrewAnnex thanks! can you run this through the auto-formatting script scripts/prepare-commit.sh ? It'll fail the travis tests due to the non-conformant source formatting.

nyalldawson commented on tests/src/providers/testqgsgdalprovider.cpp:
```diff
@@ -257,5 +259,33 @@ void TestQgsGdalProvider::bandName()
+  QgsDataProvider *provider = QgsProviderRegistry::instance()->createProvider( QStringLiteral( "gdal" ), raster, QgsDataProvider::ProviderOptions() );
+  QgsRasterDataProvider *rp = dynamic_cast< QgsRasterDataProvider * >( provider );
+  QVERIFY( rp );
+  if ( rp )
```
this can go - it's made redundant by the QVERIFY above

nyalldawson commented on tests/src/providers/testqgsgdalprovider.cpp:
```diff
@@ -257,5 +259,33 @@ void TestQgsGdalProvider::bandName()
+  QVERIFY( rp );
+  if ( rp )
+  {
+    qDebug() << "Band Name 1: " << rp->generateBandName( 1 );
```
Remove this - it's covered by QCOMPARE below, which will dump the obtained output if it differs

nyalldawson commented on tests/src/providers/testqgsgdalprovider.cpp:
```diff
@@ -257,5 +259,33 @@ void TestQgsGdalProvider::bandName()
+  QgsDataProvider *provider = QgsProviderRegistry::instance()->createProvider( QStringLiteral( "gdal" ), raster, QgsDataProvider::ProviderOptions() );
+  QgsRasterDataProvider *rp = dynamic_cast< QgsRasterDataProvider * >( provider );
+  QVERIFY( rp );
+  if ( rp )
```
same comments as above

nyalldawson commented:
>  but I noticed that the code for the DIM_EXTRA code uses a space and slash (Band 1 / NVDI). We should probably do one or the other though, I favor using the colon as it is a little more compact.

+1 to always using the colon. I think it's a better fit anyway.

This is looking good! Just some minor changes to do

AndrewAnnex committed:
addressed comments, ran prepare-commit.sh

nyalldawson commented:
Thank you!
